File: Baekjoon/gold/1238.py
from heapq import *

studentCount, lineCount, partyLocation = map(int, input().split())

toPartyLines = [[] for _ in range(studentCount + 1)]
fromPartyLines = [[] for _ in range(studentCount + 1)]
for i in range(lineCount):
    from_, to, time = map(int, input().split())
    toPartyLines[from_].append((to, time))
    fromPartyLines[to].append((from_, time))


def dijkstra(graph: list[list[tuple[int, int]]], start: int) -> list[int]:
    heapq = [(0, start)]
    time = [float("inf")] * (studentCount + 1)
    time[start] = 0

    while heapq:
        currentTime, currentLoc = heappop(heapq)
        if time[currentLoc] < currentTime:
            continue
        for nextLoc, nextTime in graph[currentLoc]:
            cost = nextTime + currentTime
            if cost < time[nextLoc]:
                time[nextLoc] = cost
                heappush(heapq, (cost, nextLoc))

    return time[1:]


maxTime = 0
for i, j in zip(
    dijkstra(toPartyLines, partyLocation), dijkstra(fromPartyLines, partyLocation)
):
    maxTime = max(i + j, maxTime)

print(maxTime)

print(dijkstra(toPartyLines, partyLocation))
print(dijkstra(fromPartyLines, partyLocation))

File: Baekjoon/gold/1238.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <cmath>
using namespace std;
const int INF = 100001;


vector<int> dijkstra(vector<pair<int, int>> graph[], int start, int studentCount) {
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> heapq;
    heapq.push({0, start});
    vector<int> time(studentCount+1, INF);
    time[start] = 0;

    while (!heapq.empty()) {
        int currentTime = heapq.top().first, currentLoc = heapq.top().second;
        heapq.pop();
        if (time[currentLoc] < currentTime) continue;

        for (pair<int, int> next : graph[currentLoc]) {
            int totalTime = currentTime + next.first;
            if (totalTime < time[next.second]) {
                time[next.second] = totalTime;
                heapq.push({totalTime, next.second});
            }
        }
    }

    return time;
}


int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int studentCount, lineCount, partyLocation;
    cin >> studentCount >> lineCount >> partyLocation;
    vector<pair<int, int>> toLine[lineCount+1];
    vector<pair<int, int>> fromLine[lineCount+1];
    for (int i = 0 ; i < lineCount ; i++) {
        int from, to, time;
        cin >> from >> to >> time;
        toLine[from].push_back({time, to});
        fromLine[to].push_back({time, from});
    }

    vector<int> toTime = dijkstra(toLine, partyLocation, studentCount);
    vector<int> fromTime = dijkstra(fromLine, partyLocation, studentCount);
    int maxTime = 0;
    for (int i = 1 ; i <= studentCount ; i++) {
        maxTime = max(maxTime, toTime[i] + fromTime[i]);
    }
    cout << maxTime;


    return 0;
}